kratos/includes/define_registry.h

```diff
@@ -0,0 +1,54 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Carlos Roig
+//                   Ruben Zorrilla
+//
+
+#pragma once
+
+// System includes
+#include <stdexcept>
+#include <sstream>
+
+
+// External includes
+
+// Project includes
+#include "includes/registry.h"
+#include "includes/registry_item.h"
+
+#define KRATOS_REGISTRY_NAME_(A,B) A##B
+#define KRATOS_REGISTRY_NAME(A,B) KRATOS_REGISTRY_NAME_(A,B)
+
+/**
+ * @brief Macro to register class prototypes
+ * This macro creates a static bool in the class (which value is not relevant and is supposed to be never accessed)
+ * This macro must be placed within the private section of the class (e.g. Process, Controller, DerivedProcess, etc)
+ * The macro forces a function to be evaluated in the declaration of the class
+ * First it checks if the provided key has been already registered (note that it checks the key but not the object)
+ * to then create a dispacher function that will initialize the registry object the first time it is accessed
+ * NOTE 1: The return type is intentionally the HasItem of the added item (and not a simple bool) in order to avoid the 
+ * compiler to statically analize the function.
+ * NOTE 2: Also note that it is explicitly needed to define the dispatcher as TFunctionType or, in other words, putting 
+ * the dispatcher function directly inside the std::move would break current code as the compiler will deduce the type
+ * of the lambda function to something that is not an std::function.
+ */
+#define KRATOS_REGISTRY_ADD_PROTOTYPE(NAME, X, ...)                                       \
+    static inline bool KRATOS_REGISTRY_NAME(_is_registered_, __LINE__) = []() -> bool {   \
+        using TFunctionType = std::function<std::shared_ptr<X>()>;                        \
+        std::string key_name = NAME + std::string(".") + std::string(#X);                 \
+        if (!Registry::HasItem(key_name))                                                 \
+        {                                                                                 \
+            auto &r_item = Registry::AddItem<RegistryItem>(key_name);                     \
+            TFunctionType dispatcher = [](){return std::make_shared<X>(__VA_ARGS__);};    \
```

Comment:
I'm a bit confused about what goes into `__VA_ARGS__` here. Are they always literals? If the aren't, they'll need to be included in the lambda's capture list, but then it won't work with literals.

Reply:
You're right. Note that so far we tried with default `Process` and `Controller`  prototypes so `__VA_ARGS__` is empty. I agree this will crash with non-default prototypes.

Reply:
No need, but they way this works (it is evaluated in the initialization of a static value) will force that all the arguments are static as well, other wise it would throw a compiler error when constructing the class. If I am not wrong that means that there is need to capture anything here.

Reply:
No, what I am saying is:

Valid: (no params)
```C++

class foo {
   ...
   KRATOS_REGISTRY_ADD_PROTOTYPE(foo)
}
```

Valid: (liters)
```C++

class foo {
   ...
   KRATOS_REGISTRY_ADD_PROTOTYPE(foo, 1, 2, "potatoe")
}
```


Valid: (statics)
```C++

class boo {
    static Parameters BooParameters;
}

boo::BooParameters = Parameters::Initialize();

class foo {
   ...
   KRATOS_REGISTRY_ADD_PROTOTYPE(foo, Parameters::BooParameters)
}
```

Valid (maybe?)
```C++

class boo {
    boo() mBooParameters(Parameters())

    Parameters mBooParameters;

    Parameters& GetParameters(return mBooParameters);
}

class foo {
   ...
   KRATOS_REGISTRY_ADD_PROTOTYPE(foo, boo().GetParameters()) // The object gets constructed INSIDE the call at execution time.
}
```

Invalid
```C++

class boo {
    boo() mBooParameters(Parameters())

    Parameters mBooParameters;

    Parameters& GetParameters(return mBooParameters);
}

global gBoo = boo();

class foo {
   ...
   KRATOS_REGISTRY_ADD_PROTOTYPE(foo, gBoo.GetParameters()) // We have no clue about the life-cycle of gBoo, may or may not exists and we would need to capture the value.
}
```

Reply:
Btw, I think this is more interesting towards the discussion of being able to directly register the `Create. For protos should suffice with the default constructor.

Reply:
That's exactly my point, I think that for our use case (i.e. registering Kratos prototypes) we want to always register with the default constructor.

About the create, I think it would be the same since, as far as I understand, we will be registering a pointer to the function that does the create from a prototype.

Long story short, for the particular case of the prototypes I'd get rid of the `__VA_ARGS__`.

Reply:
Oook